Commit php-scale-lib milestone 2

gmajor-encrypt opened the pull request:
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/vgQEZzvvFUjCyBiW6) has been filled out for this milestone. 
- [x] This pull request is being made by the same account as the accepted application. 
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application. 
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md). 

Link to the application PR: https://github.com/w3f/Grants-Program/pull/235

gmajor-encrypt commented:
@alxs Please help review

alxs commented on deliveries/php-scale-lib_milestone_2.md:
```diff
@@ -0,0 +1,17 @@
+
+**The [invoice form :pencil:](https://docs.google.com/forms/d/e/1FAIpQLSfmNYaoCgrxyhzgoKQ0ynQvnNRoTmgApz9NrMp-hd8mhIiO0A/viewform) has been filled out correctly for this milestone and the delivery is according to the official [milestone delivery guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).**
+
+* **PR Link:** [php-scale-lib](https://github.com/w3f/Open-Grants-Program/pull/235)
```
This is a bit more useful to us and conforms to the current [template](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md).

```suggestion
* **Application Document:** https://github.com/w3f/Grants-Program/blob/master/applications/php-scale-lib.md
```

alxs commented on deliveries/php-scale-lib_milestone_2.md:
```diff
@@ -0,0 +1,17 @@
+
+| Number | Deliverable | Link | Notes |
+| ------------- | ------------- | ------------- |------------- |
+| 1. | Source code   | https://github.com/gmajor-encrypt/php-scale-codec |
+| 2. | License       | MIT license |
+| 3. | Unit Tests    | https://github.com/gmajor-encrypt/php-scale-codec/tree/master/test/Codec | make test |
+| 4. | Documentation | https://github.com/gmajor-encrypt/php-scale-codec#custom-types |
+| 5. | Packagist     | https://packagist.org/packages/gmajor/substrate-codec-php |
```
Could you please structure the list of deliverables as in your [application](https://github.com/w3f/Grants-Program/blob/master/applications/php-scale-lib.md#milestone-2-example--additional-features) and provide a link to where each of them is implemented?

gmajor-encrypt commented:
@alxs Hi, these are updated, please help review

alxs commented:
Thanks @gmajor-encrypt, a few comments:

- Some tests are currently failing.
- It'd be more suitable to name the results type `Result` instead of `Results`, as the rust analogous.
- Would there be a better way to integrate the custom types into the project than importing the interfaces folder manually?
- How is the encoding of Events and Extrinsics obtained?

gmajor-encrypt commented:
> Thanks @gmajor-encrypt, a few comments:
>
> * Some tests are currently failing.
> * It'd be more suitable to name the results type `Result` instead of `Results`, as the rust analogous.
> * Would there be a better way to integrate the custom types into the project than importing the interfaces folder manually?
> * How is the encoding of Events and Extrinsics obtained?

Thanks for review
1.  Please run ```composer install``` again
2. Fixed
3. There are many ways to import custom type, just change the imported type to array. Similar to readme https://github.com/gmajor-encrypt/php-scale-codec#custom-types
4. m2 does not include encoding of Events and Extrinsics, if you test, you can do this

> get block extrinsics
```bash
websocat wss://polkadot.api.onfinality.io/public-ws
{"id":1,"method":"chain_getBlock","params":["block_hash"],"jsonrpc":"2.0"}
```
> get block event record
```bash
websocat wss://polkadot.api.onfinality.io/public-ws
{"id":1,"method":"state_getStorageAt","params":["0x26aa394eea5630e07c48ae0c9558cef780d41e5e16056765bc8461851072c9d7",block_hash"],"jsonrpc":"2.0"}
```
``0x26aa394eea5630e07c48ae0c9558cef780d41e5e16056765bc8461851072c9d7`` is eventRecord storageKey

The method of encoding Extrinsics is currently useless in scale codec, RPC request is needed here, I am going to develop in libs of PHP RPC API https://github.com/gmajor-encrypt/php-substrate-api

alxs commented:
Thanks! Feel free to add the dependencies to the README.

Besides, could you add some instructions on how to use the types in https://github.com/gmajor-encrypt/php-scale-codec/tree/master/src/Codec/interfaces? And do you know how other implementations handle these types? Manually extracting the JSON data from the files in @polkadot/api works, but is arguably not very elegant.

Thanks for the instructions, that's exactly what I was looking for. However I think there's a missing step - I'm not familiar with `websocat` and simply connecting to the socket and then entering the second line of your examples returns

```
{"jsonrpc":"2.0","error":{"code":-32602,"message":"Invalid params: 0x prefix is missing."},"id":1}
```

gmajor-encrypt commented:
@alxs , thank you reply

At present, these types are extracted from https://github.com/paritytech/substrate. @polkadot/api, as the most complete third-party library, is currently the most complete. I also refer to https://github.com/itering/scale.go of subscan, this method is not very elegant, but it is very practical.

In the latest https://github.com/paritytech/substrate/pull/8615, all types can be obtained from metadata, but this has not been released yet, and it may take some time.

For now, this solution is sufficient, and many third-party libs are in this way (https://github.com/centrifuge/go-substrate-rpc-client uses a static solution, this way every time All need to modify the source code, can not rely on file registration)

gmajor-encrypt commented:
I don’t know if you noticed that there is a block_hash in params. This is a parameter and needs to be passed in

alxs commented:
- Agree that although the method is perhaps not the most elegant, it's likely the way to go for the moment :+1: 
- Is there really [no difference](https://github.com/gmajor-encrypt/php-scale-codec/blob/master/src/Codec/Types/metadataV13.php) between the decoding of V12 and V13 metadata? Is that class even [used](https://github.com/gmajor-encrypt/php-scale-codec/blob/master/src/Codec/Types/Metadata.php#L16)?
- A reminder to add some user documentation on custom types, please provide more than a few examples for reference https://github.com/gmajor-encrypt/php-scale-codec#custom-types You can also link to existing documentation if possible. But I currently have no way of knowing what types are supported or what the syntax is. The `asserts` are also irrelevant in that context.
- Could you also provide user documentation on the format of the decoded Metadata, Extrinsic and Event types?
- Could you implement some meaningful tests for metadata? Currently the [only thing tested](https://github.com/gmajor-encrypt/php-scale-codec/blob/master/test/Codec/Test/MetadataTest.php) is that a) the parser doesn't crash and b) the metadata version is correct, and only one test case exists for each version. Same goes for [extrinsics](https://github.com/gmajor-encrypt/php-scale-codec/blob/master/test/Codec/Test/ExtrinsicTest.php), [events](https://github.com/gmajor-encrypt/php-scale-codec/blob/master/test/Codec/Test/EventTest.php) and [custom types](https://github.com/gmajor-encrypt/php-scale-codec/blob/master/test/Codec/Test/BaseTest.php), for each of which only 1-2 cases are tested.
- A reminder that code documentation is also required as per our [guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md#documentation). The project currently only contains a few auto-generated DocBlocks which don't really help its readability and would benefit a lot from comprehensive documentation. I previously [brought this up](https://github.com/w3f/Grant-Milestone-Delivery/pull/105#issuecomment-804635051) for M1.

alxs reviewed (changes requested):
- Agree that although the method is perhaps not the most elegant, it's likely the way to go for the moment :+1: 
- Is there really [no difference](https://github.com/gmajor-encrypt/php-scale-codec/blob/master/src/Codec/Types/metadataV13.php) between the encoding of V12 and V13 metadata?
- A reminder to add some user documentation on custom types, please provide more than a few examples for reference https://github.com/gmajor-encrypt/php-scale-codec#custom-types. You can also link to existing documentation if possible. But I currently have no way of knowing what types are supported or what the syntax is. The `asserts` are also irrelevant in that context.
- Could you also provide user documentation on the format of the decoded Metadata, Extrinsic and Event types?
- Could you implement some meaningful tests for metadata? Currently the [only thing tested](https://github.com/gmajor-encrypt/php-scale-codec/blob/master/test/Codec/Test/MetadataTest.php) is that a) the parser doesn't crash and b) the metadata version is correct, and only one test case exists for each version. Same goes for [extrinsics](https://github.com/gmajor-encrypt/php-scale-codec/blob/master/test/Codec/Test/ExtrinsicTest.php), [events](https://github.com/gmajor-encrypt/php-scale-codec/blob/master/test/Codec/Test/EventTest.php) and [custom types](https://github.com/gmajor-encrypt/php-scale-codec/blob/master/test/Codec/Test/BaseTest.php), for each of which only 1-2 cases are tested.
- A reminder that code documentation is also required as per our [guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md#documentation). The project currently only contains a few auto-generated DocBlocks which don't really help its readability and would benefit a lot from comprehensive documentation. I previously [brought this up](https://github.com/w3f/Grant-Milestone-Delivery/pull/105#issuecomment-804635051) for M1.

gmajor-encrypt commented:
@alxs Some updates have been made, please help to review

alxs commented:
Thanks @gmajor-encrypt. I quickly checked the parsed metadata and, while the parsed object seems correct and the metadata tests in polkadot-js also only test one instance, at least the entire parsed object is verified against its JSON counterpart. Do you think you could do the same?

Apart from that, you haven't addressed most of the points in my last comment and seeing as this evaluation has also been going on for quite a while now, I would like to ask you to please make sure to do so before submitting another update to us and please within the next 2 weeks, else we will need to terminate the grant due to it failing to meet the delivery standards and being overdue.

Since you mentioned in chat implementing metadata v14 in the first week of October which is right around the corner, I'd also be happy to give you 1-2 more weeks if you want to implement it and have it reviewed within this delivery.

gmajor-encrypt commented:
@alxs Hey, Metadata parse has refactor, added metadata encode, test as been update https://github.com/gmajor-encrypt/php-scale-codec/blob/master/test/Codec/Test/MetadataTest.php#L51

gmajor-encrypt commented:
For custom type, I added more documentation https://github.com/gmajor-encrypt/php-scale-codec#custom-types
About the syntax of custom types https://github.com/gmajor-encrypt/php-scale-codec/blob/master/custom_type.md
For point 4, https://github.com/gmajor-encrypt/php-scale-codec#metadata, as well as the following extrince and event, there are corresponding structures and descriptions of links to external documents in the readme and code.
For point6, I have added descriptions and comments for all necessary code blocks

alxs commented:
Hey, thanks for the updates.

The [newly added documentation](https://github.com/gmajor-encrypt/php-scale-codec/commit/901e71ef9f4748176d9188352c4da81cf35c7b42) for Metadata looks great. However, it of course makes little sense for only one type to be well-documented; I was referring to the project in general. Could you add such documentation project-wide?

I also noticed that Metadata v14 support is the only deliverable pertaining to this project in your new application. As we'd rather have the projects separated across grants, I'd rather ask you to submit an amendment to include it in this milestone. However, I'd also like to point out that you [committed to keeping the project](https://github.com/w3f/Grants-Program/pull/235#issuecomment-774009557) up to date for 1 year after release as part of this grant.

gmajor-encrypt commented:
@alxs Thank you reply.

Regarding metadata v14, do you mean that you want to complete it at this milestone? Because you mentioned above that the remaining time is not in a hurry, can I make a new milestone, because there are many things involved, because Polkadot and Kusama have not been updated to v14, and some tests are missing, there may be some errors here,

In addition, I did promise to keep updated within a year, but since this milestone has not been completed, I am the first to complete the content needed for this milestone

alxs commented:
What I meant is that it was agreed maintenance would be included in the total cost. Hence it shouldn't incur an increase in price or constitute a deliverable in a follow-up grant, but if you'd still like it to be evaluated you can submit it with this milestone (for which actually no amendment would be needed) or as a separate milestone (for which you'd need to submit an amendment).

gmajor-encrypt commented:
I am preparing to submit m2. I agree with your suggestion. I will develop v14 after the end of m2. It is not included in any milestones. It is only the maintenance of this project.

gmajor-encrypt commented:
Ok, I will complete the development of metadata v14 in m2. You mentioned above that he is not within the scope of evaluation. I hope to give me another 2-3 weeks to complete it.

gmajor-encrypt commented:
> @gmajor-encrypt sounds good! Could you finish the documentation first? Then I'll put the evaluation on hold while you work on metadata v14.

Hey, alex, Yes, I have basically completed all the documents. Please help me to check it. I will start the development of v14 the day after tomorrow.
